Add OrchardCore.Form client side and server side Validate Rule Function

SpikeXy opened the pull request:
client side validate use this js : 
https://github.com/validatorjs/validator.js

only input and textArea support：

![image](https://user-images.githubusercontent.com/12979622/108737165-4a197700-756d-11eb-900d-259e41141c8b.png)


Reference documents：
https://github.com/OrchardCMS/OrchardCore/issues/3980

https://docs.orchardcore.net/en/dev/docs/reference/modules/Forms/#adding-name-email-and-message-validation

Piedone commented on src/OrchardCore.Modules/OrchardCore.Forms/Entities/Scripting/ValidateRuleMethod.cs:
```diff
@@ -0,0 +1,144 @@
+
+namespace OrchardCore.Forms.Entities.Scripting
+{
+    public class ValidateRuleMethod : IGlobalMethodProvider
```
There are a couple of small things to fix: 
![image](https://user-images.githubusercontent.com/1976647/109736063-1de0a480-7bc4-11eb-94de-2adef0249c43.png)

Also, run Visual Studio formatting (Ctrl+K+D) on the file.

Piedone replied:
The whole class if very impressive!

Piedone commented on src/OrchardCore.Modules/OrchardCore.Forms/Migrations.cs:
```diff
@@ -53,6 +53,12 @@ public int Create()
                 .WithPart("LabelPart")
                 .Stereotype("Widget"));
 
+
+
+            //ValidationRule
```
```suggestion

            // ValidationRule
```

Piedone commented on src/OrchardCore.Modules/OrchardCore.Forms/Migrations.cs:
```diff
@@ -53,6 +53,12 @@ public int Create()
+
+            //ValidationRule
+            _contentDefinitionManager.AlterPartDefinition("ValidationRulePart", part => part
+                .WithDescription("Provides Validate Rule property."));
```
```suggestion
                .WithDescription("Provides the ability to use and configure validation rules for form fields.."));
```

Piedone commented on src/OrchardCore.Modules/OrchardCore.Forms/Migrations.cs:
```diff
@@ -53,6 +53,12 @@ public int Create()
                 .WithPart("LabelPart")
                 .Stereotype("Widget"));
 
+
```
Add these to an `UpdateFrom3()` method too so existing installations will get them as well.

Piedone commented on src/OrchardCore.Modules/OrchardCore.Forms/Views/Form.Wrapper.cshtml:
```diff
@@ -7,16 +7,27 @@
+<script at="Foot">
+    var formId = '#'+ '@formId'
+    $(formId).on('click', function () {
+        console.log(window.validatorObject)
```
Leftover debug logging.

Piedone commented on src/OrchardCore.Modules/OrchardCore.Forms/Views/Form.Wrapper.cshtml:
```diff
@@ -7,16 +7,27 @@
+    var formId = '#'+ '@formId'
+    $(formId).on('click', function () {
+        console.log(window.validatorObject)
+        $("<input />").attr("type", "hidden")
```
```suggestion
        $("<input />")
            .attr("type", "hidden")
```

Piedone commented on src/OrchardCore.Modules/OrchardCore.Forms/Views/Form.Wrapper.cshtml:
```diff
@@ -7,16 +7,27 @@
 </form>
+<script at="Foot">
+    var formId = '#'+ '@formId'
+    $(formId).on('click', function () {
```
Why not do this on document.ready?

Piedone commented on src/OrchardCore.Modules/OrchardCore.Forms/Views/Form.Wrapper.cshtml:
```diff
@@ -7,16 +7,27 @@
 
     @await DisplayAsync(Model.Metadata.ChildContent)
 </form>
+<script at="Foot">
```
Leave a blank line before it.

Piedone commented on src/OrchardCore.Modules/OrchardCore.Forms/Views/Items/ValidateRuleFormTask.Fields.Thumbnail.cshtml:
```diff
@@ -0,0 +1,2 @@
+<h4 class="card-title"><i class="fa fa-shield-alt"></i>@T["Validate Rule Form"]</h4>
+<p>@T["Checks whether the submitted form contains any validation Rule errors."]</p>
```
```suggestion
<p>@T["Checks whether the submitted form contains any validation rule errors."]</p>
```

Piedone commented on src/OrchardCore.Modules/OrchardCore.Forms/Views/Items/ValidationRulePart.Fields.Edit.cshtml:
```diff
@@ -0,0 +1,27 @@
+@model ValidationRulePartEditViewModel
+
+<div class="form-group" >
+    <label asp-for="Type">@T["Validate Type"]</label>
```
The expression is more like "validation" than "validate" so please change it in every text.

Piedone commented on src/OrchardCore.Modules/OrchardCore.Forms/Views/Items/ValidationRulePart.Fields.Edit.cshtml:
```diff
@@ -0,0 +1,27 @@
+</div>
+
+<div class="form-group" >
+    <label asp-for="Option">@T["Validate Options"]</label>
+    <input asp-for="Option" type="text" class="form-control content-preview-text" />
+    <span class="hint">@T["The text to check the string. For detailed rules, please see https://github.com/validatorjs/validator.js"]</span>
```
Could you add some docs on using this? https://docs.orchardcore.net/en/dev/docs/topics/docs-contributions/

Also, make the GitHub URL a link.

Piedone replied:
Can you explain "The text to check the string."? I don't really understand.

Piedone commented on src/OrchardCore.Modules/OrchardCore.Forms/Views/Items/ValidationRulePart.cshtml:
```diff
@@ -0,0 +1,64 @@
+
+
+        $(elementId).bind("keyup", function (e) {
+            var re = false;
```
What does `re` mean?

Piedone replied:
What does `re` mean?

Piedone commented on src/OrchardCore.Modules/OrchardCore.Forms/Views/Items/ValidationRulePart.cshtml:
```diff
@@ -0,0 +1,64 @@
+            var docEle = document.getElementById(validateElementId);
+            if (re) {
+                if (docEle) {
+                    $(elementId).parent().parent().parent().find(":submit").removeAttr('disabled');
```
Maybe find the submit button recursively instead of hard-coding a depth? Below too.

Piedone commented on src/OrchardCore.Modules/OrchardCore.Forms/Views/Items/ValidationRulePart.cshtml:
```diff
@@ -0,0 +1,64 @@
+               re = window.validator['@type']($(elementId).val(), JSON.parse(option));
+            }
+
+            var docEle = document.getElementById(validateElementId);
```
Don't abbreviate variables. Maybe `validationElement`?

Piedone commented on src/OrchardCore.Modules/OrchardCore.Forms/Views/Items/ValidationRulePart.cshtml:
```diff
@@ -0,0 +1,64 @@
+            obj.option = option;
+            objs.push(obj);
+            window.validatorObject.push(obj);
+        } else {
```
```suggestion
        }
        else {
```

Below too.

Piedone commented on src/OrchardCore.Modules/OrchardCore.Forms/Views/Items/ValidationRulePart.cshtml:
```diff
@@ -0,0 +1,64 @@
+                }
+            } else {
+                if (!docEle) {
+                    var dangerText = "<p id='" + validateElementId + "' class='text-danger'>Validate " + "@type" + " failed.<p>";
```
The text needs to be localized.

Piedone commented on src/OrchardCore.Modules/OrchardCore.Forms/Models/ValidationRulePart .cs:
```diff
@@ -0,0 +1,12 @@
+        public bool IsNeedValidate { get; set; }
+        public string Type { get; set; }
+        public string Option { get; set; }
+
```
Unnecessary line break.

Piedone commented on src/OrchardCore.Modules/OrchardCore.Forms/Models/ValidationRulePart .cs:
```diff
@@ -0,0 +1,12 @@
+{
+    public class ValidationRulePart : ContentPart
+    {
+        public bool IsNeedValidate { get; set; }
```
Unused property.

Piedone commented on src/OrchardCore.Modules/OrchardCore.Forms/Entities/Scripting/ValidateRuleMethod.cs:
```diff
@@ -0,0 +1,144 @@
+using System;
```
Can you include a recipe (just attach under the PR) that contains a demonstration of how to use this, with a form?

Piedone commented on src/OrchardCore.Modules/OrchardCore.Forms/Entities/Scripting/ValidateRuleMethod.cs:
```diff
@@ -0,0 +1,144 @@
+using System;
```
Can you include a recipe (just attach under the PR) that contains a demonstration of how to use this, with a form?

SpikeXy replied:
https://gist.github.com/SpikeXy/8d9dc39fa3cde80076e5f9883c3e43f0, the workflow does not created when the recipe inserted, maybe something wrong, so if we want to test methods in ValidateRuleMethod.cs we need to create the workflow manually.

Piedone replied:
Thank you!

Piedone commented:
Please don't resolve conversations, let me do that so I can check how you addressed my remarks. Let me know when you're done.

SpikeXy commented:
> Please don't resolve conversations, let me do that so I can check how you addressed my remarks. Let me know when you're done.

Ok, I understand the process, and now I have completed it. Can I start to modify this document after you check it?
https://docs.orchardcore.net/en/ag-docs/docs/reference/modules/Forms/

Piedone commented:
Sure, you can include docs modifications already (it's in the src/docs folder), see: https://docs.orchardcore.net/en/dev/docs/topics/docs-contributions/

SpikeXy commented:
> Sure, you can include docs modifications already (it's in the src/docs folder), see: https://docs.orchardcore.net/en/dev/docs/topics/docs-contributions/

Ok i will start right away.

Piedone commented:
Let me know if you're done here.

SpikeXy commented:
I've done it.

Piedone commented:
Ah OK. Next time just use the "request review" feature under the PRs of GitHub or just let me know in a comment, as I asked for, otherwise I won't know when you're done :).

Piedone commented:
Also, please don't resolve pull request conversations while addressing them, I'd do that so I can check the prior feedback. However, you can keep track that you've addressed something with a reaction under the message for example.

Piedone commented on src/OrchardCore.Modules/OrchardCore.Forms/Assets.json:
```diff
@@ -14,7 +14,7 @@
   {
     "inputs": [
-      "Assets/js/SelectOptionsEditor/validator.js"
+      "Assets/js/Scripts/validator.js"
```
This is still not the correct path.

Piedone replied:
This is still there. Or do you disagree?

Piedone commented on src/OrchardCore.Modules/OrchardCore.Forms/Migrations.cs:
```diff
@@ -149,6 +141,28 @@ public int UpdateFrom2()
         }
+        public int UpdateFrom3()
+        {
+            //ValidationRule
```
```suggestion
            // ValidationRule
```

Piedone commented on src/OrchardCore.Modules/OrchardCore.Forms/Migrations.cs:
```diff
@@ -149,6 +141,28 @@ public int UpdateFrom2()
 
             return 3;
         }
+        public int UpdateFrom3()
+        {
+            //ValidationRule
+            _contentDefinitionManager.AlterPartDefinition("ValidationRulePart", part => part
+                .WithDescription("Provides the ability to use and configure validation rules for form fields.")
+            );
+
+
+            _contentDefinitionManager.AlterTypeDefinition("Input", type => type
+                .WithPart("ValidationRulePart", part => part
+                    .WithPosition("3")
+                 )
+             );
+
+            _contentDefinitionManager.AlterTypeDefinition("TextArea", type => type
+                .WithPart("ValidationRulePart", part => part
+                    .WithPosition("3")
+                )
+            );
+
+            return 4;
+        }
```
This is good like this but you also need to add the same migration logic back to `Create()` too so fresh installs don't have to execute the separate update step. Here it is explained why this is useful: https://github.com/Lombiq/Orchard-Training-Demo-Module/blob/dev/Migrations/BookMigrations.cs#L46

Piedone replied:
`ValidationRulePart`'s description being added is missing from `Create()`.

Piedone commented on src/OrchardCore.Modules/OrchardCore.Forms/Views/Form.Wrapper.cshtml:
```diff
@@ -20,14 +20,17 @@
-            .attr("value", JSON.stringify(window.validatorObject))
-            .appendTo(formId);
-        return true;
+    $(function () {
```
It's best to use the `jQuery` variable directly instead of the dollar sign because the latter can be overloaded. You can wrap this whole block into a 
```js
(function ($) {
...
})(jQuery);
```
to work around this. See: https://orcharddojo.net/orchard-resources/CoreLibrary/DevelopmentGuidelines/BestPractices/JavaScript

Piedone commented on src/OrchardCore.Modules/OrchardCore.Forms/Views/Items/ValidationRulePart.Fields.Edit.cshtml:
```diff
@@ -2,26 +2,33 @@
+    <label asp-for="Option">@T["Validation Options"]</label>
     <input asp-for="Option" type="text" class="form-control content-preview-text" />
-    <span class="hint">@T["The text to check the string. For detailed rules, please see https://github.com/validatorjs/validator.js"]</span>
+    <span class="hint">@T["For detailed, please visit https://github.com/validatorjs/validator.js"]</span>
```
Please make the GitHub URL into an HTML link.

Also:
```suggestion
    <span class="hint">@T["For details, please visit https://github.com/validatorjs/validator.js"]</span>
```

Piedone commented on src/OrchardCore.Modules/OrchardCore.Forms/Views/Items/ValidationRulePart.cshtml:
```diff
@@ -3,62 +3,56 @@
+                    var dangerText = "<div id='" + validateElementId + "' class='text-danger'>" + validateStr + "</div>";
                     $(elementId).after(dangerText);
-                    $(elementId).parent().parent().parent().find(":submit").attr('disabled', 'true');
+                    $(elementId).closest("form").find(":submit").attr('disabled', 'true');
```
Use quotes for strings consistently. Better use single quotes (`'`) everywhere.

Piedone commented on src/OrchardCore.Modules/OrchardCore.Forms/Views/Items/ValidationRulePart.cshtml:
```diff
@@ -3,62 +3,56 @@
+            }
+            else {
+                if (!validationElement) {
+                    var validateStr =  '@T[@type]' + ' @T["validated failed"]';
+                    var dangerText = "<div id='" + validateElementId + "' class='text-danger'>" + validateStr + "</div>";
```
Use params in localized string for dynamic values:
```suggestion
                    var validationMessage =  '@T["{0} validation failed", type]';
                    var dangerText = '<div id="' + validateElementId + '" class="text-danger">' + validationMessage + '</div>';
```

Piedone commented on src/OrchardCore.Modules/OrchardCore.Forms/Views/Items/ValidationRulePart.cshtml:
```diff
@@ -3,62 +3,56 @@
+    function validateElementFunction(type, originElementId, option) {
+        var elementId = '#' + originElementId;
+        var timestamp = new Date().getTime();
+        var validateElementId = originElementId + timestamp;
```
Maybe?

```suggestion
        var validationElementId = originElementId + timestamp;
```

Piedone commented on src/docs/reference/modules/Forms/README.md:
```diff
@@ -215,6 +215,40 @@ The final workflow should look like this:
 
 ![Contact Form Workflow](./assets/contact-form-workflow-1.png)
 
+## Validate Rule Usage
+validate plug : [validator.js](https://github.com/validatorjs/validator.js)
+Validate rule is a validate option only used by input widget and textarea widget.It can validate input in frontend and backend.Now these functions are implemented:
+
+|Validator|Description|Options|
+|--|--|--|
+|None|don't validate value|none|
+|Contains|check if the string contains the seed( case sensitive always) | compare string |
+|Equals|check if the string matches the comparison( no case sensitive always)|compare string |
+|IsAfter|check if the string is a date that's after the specified date(needed).| datetime,like 2020-3-3 |
+|IsBefore|check if the string is a date that's before the specified date(needed).| datetime,like 2020-3-3 |
+|IsBoolean|check if a string is a boolean.|none|
+|IsByteLength|check if the string's length (in UTF-8 bytes) falls in a range.| {"min":0, "max": 20} |
+|IsDate|Check if the input is a valid date|same as validator.js|
+|IsDecimal|check if the string represents a decimal number, such as 0.1, .3, 1.1, 1.00003, 4.0, etc.|same as validator.js|
+|IsDivisibleBy|check if the string is a number that's divisible by another.|Divisor,like 3
+|IsEmpty|check if the string has a length of zero.|same as validator.js.|
+|IsFloat|check if the string is a float.|{ "min": 7.22, "max": 9.55 } |
+|IsInt|check if the string is an integer.|same as validator.js|
+|IsJSON|check if the string is valid JSON (note: uses JSON.parse).|same as validator.js|
+|IsLength|check if the string's length falls in a range.|{ "min": 7.22, "max": 20.0 }|
+|IsNumeric|check if the string contains only numbers.|same as validator.js|
+|Matches|check if string matches the pattern.|regex pattern string like : ^\\\\d{n}$,double backslash need if you want to input single backslash |
+
+
+It can seen from below pic,the submit button will be available only without validate rule errors.
+![Form Frontend Use Validate Rule](./assets/form-validation-rule-frontend.png)
+
+
+if you want to use validate rule in backend, you should create validate rule workflow like this:
+![Form Backend Use Validate Rule Workflow](./assets/form-validation-rule-backend.png)
```
```suggestion
## Validation Rule Usage

This feature uses [validator.js](https://github.com/validatorjs/validator.js) to offer a variety of validation rules for form fields. They can only be used by the Input Widget and TextArea Widget. Validating input on the frontend and backend are both supported.

Currently, the following functions are implemented:

|Validator|Description|Options|
|--|--|--|
|None|don't validate value|none|
|Contains|Check if the string contains the seed (always case-sensitive). | compared string |
|Equals|Check if the string matches the comparison (alway case-insensitive)|compared string |
|IsAfter|Check if the string is a date that's after the specified date(needed).| datetime, like `2020-3-3` |
|IsBefore|Check if the string is a date that's before the specified date(needed).| datetime, like `2020-3-3` |
|IsBoolean|Check if a string is a boolean.|none|
|IsByteLength|Check if the string's length (in UTF-8 bytes) falls within a range.| `{"min":0, "max": 20}` |
|IsDate|Check if the input is a valid date|Same as validator.js.|
|IsDecimal|Check if the string represents a decimal number, such as 0.1, .3, 1.1, 1.00003, 4.0, etc.|Same as validator.js.|
|IsDivisibleBy|Check if the string is a number that's divisible by another.|Divisor, like `3`.
|IsEmpty|Check if the string has a length of zero.|Same as validator.js.|
|IsFloat|Check if the string is a float.|`{ "min": 7.22, "max": 9.55 }`|
|IsInt|Check if the string is an integer.|Same as validator.js.|
|IsJSON|Check if the string is valid JSON (note: uses `JSON.parse()`).|Same as validator.js.|
|IsLength|Check if the string's length falls in a range.|`{ "min": 7.22, "max": 20.0 }`|
|IsNumeric|Check if the string contains only numbers.|Same as validator.js.|
|Matches|Check if string matches the pattern.|Regex pattern string like: `^\\\\d{n}$`, double backslash needed if you want to input single backslash. |


As can be seen from the below picture, the submit button will be available only if there are no validation rule errors.
![Form Frontend Use Validation Rule](./assets/form-validation-rule-frontend.png)


if you want to use validation rules in the backend, you should create a validation rule workflow like this:
![Form Backend Use Validation Rule Workflow](./assets/form-validation-rule-backend.png)
```

Piedone reviewed (changes requested):
Thank you, awesome! Would you like to demo it during the next Orchard meeting (Tuesdays at 12 PM PT)? Or I can demo it, explaining that it's your work.

SpikeXy commented:
I have fixed suggestion.It would be great if you can demo it, my listening can’t keep up with your speaking speed at the moment.

Piedone commented on src/OrchardCore.Modules/OrchardCore.Forms/Entities/Scripting/ValidateRuleMethod.cs:
```diff
@@ -1,142 +1,249 @@
+                Method = serviceProvider => (Func<string, bool>)((str) =>
+                {
+                    bool flag;
+                    return Boolean.TryParse(str, out flag);
```
There are other similar cases in the class.

Piedone commented on src/OrchardCore.Modules/OrchardCore.Forms/Views/Items/ValidationRulePart.Fields.Edit.cshtml:
```diff
@@ -29,6 +29,6 @@
 <div class="form-group" >
     <label asp-for="Option">@T["Validation Options"]</label>
     <input asp-for="Option" type="text" class="form-control content-preview-text" />
-    <span class="hint">@T["For detailed, please visit https://github.com/validatorjs/validator.js"]</span>
```
You can include the `a` tag in the localized string :).

Piedone commented on src/OrchardCore.Modules/OrchardCore.Forms/Entities/Scripting/ValidateRuleMethod.cs:
```diff
@@ -1,142 +1,249 @@
 using System.Globalization;
-using System.Linq;
-using System.Net;
+using System.Text;
```
It still could use it :). There are formatting issues in the file.

SpikeXy replied:
[https://github.com/SpikeXy/OrchardCore/blob/dev/src/OrchardCore.Modules/OrchardCore.Forms/Entities/Scripting/ValidateRuleMethod.cs](https://github.com/SpikeXy/OrchardCore/blob/dev/src/OrchardCore.Modules/OrchardCore.Forms/Entities/Scripting/ValidateRuleMethod.cs)

The using code has been modified in pull request, can you see it ?

Piedone replied:
If you do a format in Visual Studio you'll see something like this:
![image](https://user-images.githubusercontent.com/1976647/113286781-f5e08000-92ec-11eb-8e3b-bda03c60d8e8.png)

SpikeXy replied:
ok,done.

Piedone reviewed (changes requested):
Sure then! Would you like me to demo it at next week's Tuesday meeting? Would you like to join just to simply be there (you don't have to follow, if that makes sense)?

SpikeXy commented:
Of course I would. It's my pleasure.

Piedone commented:
Great then! Let's get everything done here then.

Piedone commented:
Awesome! Now let's do the following.

1. Please open this same PR in the OC repo here: https://github.com/OrchardCMS/OrchardCore/pulls Use the title "Rule-based validation of form elements (Lombiq Technologies: OCORE-51)", please.
2. I'll demo it during the next meeting, at 12 PM PT, here: https://orchardcore.net/meeting I'll explain that it's your work, I'm just demoing it. I'll also relay any feedback from the meeting under the PR.
3. I'm closing this PR, not as a rejection but just because we shouldn't merge it here but in the OC repo. But I don't have anything else to add here, thank you for your work!

Piedone commented:
Sorry Spike, I just figured that the Workflow is not working for me anymore (just get a 404 on form submission) and I couldn't get it quickly working so I won't be able to demo today. Would you please upload an updated recipe that you can confirm works?

SpikeXy commented:
[https://gist.github.com/SpikeXy/8d9dc39fa3cde80076e5f9883c3e43f0](https://gist.github.com/SpikeXy/8d9dc39fa3cde80076e5f9883c3e43f0)，import this recipe,then turn to workflow 'testform',click edit into 'testFormRequest',then click Regenerate , copy this link and click save.Finally change page 'testFormPage' and 'testMatchPage' action link by copied link.This will work.

Piedone commented:
Thank you!

I don't see validation working though (this has worked at an earlier stage), should this work?

https://user-images.githubusercontent.com/1976647/113947152-46367100-980a-11eb-885c-97b576f85d75.mp4

SpikeXy commented:
It's working.NoneValidate will not validate anything,containsValidate execution is right.Because 'thinkd' contains 'think'.But this submission confuses you,because i don't set a not valid workflow redirect link.
[https://gist.github.com/SpikeXy/8d9dc39fa3cde80076e5f9883c3e43f0](https://gist.github.com/SpikeXy/8d9dc39fa3cde80076e5f9883c3e43f0)
Maybe here need submit onclick check event, it will disable submit if you have something error like leave blank.

Piedone commented:
Ah I see now, thank you. I have two new remarks though, let me reopen this.

Piedone commented on src/OrchardCore.Modules/OrchardCore.Forms/Views/Items/ValidationRulePart.cshtml:
```diff
@@ -0,0 +1,58 @@
+            }
+            else {
+                if (!validationElement) {
+                    var validationMessage =  '@T["{0} validation failed", type]';
```
Just a nitpick: Since this looks like this:

![image](https://user-images.githubusercontent.com/1976647/114028761-1ecabd00-9879-11eb-9d17-a2981dc2981f.png)

I'd update it to make it more like a proper sentence:
```suggestion
                    var validationMessage =  '@T["Validation failed for \"{0}\".", type]';
```

Piedone replied:
Could this be just a default, and allow you to add a custom validation message when configuring the given rule, i.e. in the form editor?

Piedone commented on src/OrchardCore.Modules/OrchardCore.Forms/Views/Items/ValidationRulePart.cshtml:
```diff
@@ -0,0 +1,58 @@
+<script asp-src="~/OrchardCore.Forms/Scripts/validator.min.js" debug-src="~/OrchardCore.Forms/Scripts/validator.js" asp-name="validator" depends-on="jQuery" at="Foot"></script>
+
+<script at="Foot">
+    function validateElementFunction(type, originElementId, option) {
```
Client-side validation errors will only appear if you've previously edited the given field:

https://user-images.githubusercontent.com/1976647/114029067-71a47480-9879-11eb-8c69-d2772d10be4c.mp4

I think they should appear when you try to submit the form in also if you didn't edit them.

SpikeXy commented:
Finished, this is new recipe: [https://gist.github.com/SpikeXy/8d9dc39fa3cde80076e5f9883c3e43f0](https://gist.github.com/SpikeXy/8d9dc39fa3cde80076e5f9883c3e43f0)

Piedone commented on src/OrchardCore.Modules/OrchardCore.Forms/Views/Items/ValidationRulePart.cshtml:
```diff
@@ -24,35 +64,23 @@
     }
     if ('@type' != 'none') {
-        validateElementFunction('@type', '@elementId', '@option');
+        validateElementFunction('@type', '@elementId', '@option','@validationMessage');
```
If you fix one of the validation errors then you'll be able to submit the form without further client-side validation errors. Server-side validation will work but I think you should still get client-side validation there.

https://user-images.githubusercontent.com/1976647/114313815-6743de00-9af8-11eb-8678-89229bf0c844.mp4

Piedone commented on src/OrchardCore.Modules/OrchardCore.Forms/Views/Items/ValidationRulePart.cshtml:
```diff
@@ -3,17 +3,57 @@
 
 <script asp-src="~/OrchardCore.Forms/Scripts/validator.min.js" debug-src="~/OrchardCore.Forms/Scripts/validator.js" asp-name="validator" depends-on="jQuery" at="Foot"></script>
 
 <script at="Foot">
```
This whole block is repeated for every form field. Rather, only the `validateElementFunction()` call should.

Piedone reviewed (changes requested):
I noticed a couple of new issues. But nevertheless, I'll demo on the Tuesday meeting.

I've updated the recipe with a few fixes to the demo forms and added validation error/success pages.
[Demo.zip](https://github.com/Lombiq/OrchardCore/files/6292409/Demo.zip)

SpikeXy commented:
Fixed, and i updated recipe gist.This gist added success and error page. [https://gist.github.com/SpikeXy/8d9dc39fa3cde80076e5f9883c3e43f0](https://gist.github.com/SpikeXy/8d9dc39fa3cde80076e5f9883c3e43f0)

Piedone commented on src/OrchardCore.Modules/OrchardCore.Forms/Assets.json:
```diff
@@ -17,5 +17,11 @@
+  },
+  {
+    "inputs": [
+      "Assets/js/Validate/validationRuleHelper.js"
```
Sorry, I just noticed the folder name. It should also rather be "Validation".

Piedone reviewed (changes requested):
Thank you! Will demo tomorrow. It's at 12 PM PT under https://orchardcore.net/meeting. I'm curious what the others from the community will say.

SpikeXy commented:
Ok,The folder name has changed.^_^

Piedone commented:
Thank you, perfect!

Piedone commented:
Thanks again! Please now open a PR in the OC repo now.

Some feedback from the meeting:

- It's not clear how Validation Options should be used. At least automatically populating it with an example for the selected rule (or adding a placeholder text to the input field for it) would help a lot.
- `ValidateRuleMethod` could directly execute the JS library or just execute the server-side validation directly in C# and not as `GlobalMethod`. If we use JS scripting then a validator could be to use free JS code too.
